scripts/4.Covariates.py

```diff
@@ -0,0 +1,222 @@
+    n_feature_kept = 500
+    param_fixed = {
+        'loss': 'log',
+        'penalty': 'elasticnet',
+    }
+    param_grid = {
+        'alpha': [10 ** x for x in range(-6, 1)],
+        'l1_ratio': [0, 0.05, 0.1, 0.2, 0.5, 0.8, 0.9, 0.95, 1],
+    }
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
+
+    # select the top features with the highest MAD
+    feature_select = SelectKBest(fs_mad, k=n_feature_kept)
+
+    # Include loss='log' in param_grid doesn't work with pipeline somehow
+    clf = SGDClassifier(random_state=0, class_weight='balanced',
+                        loss=param_fixed['loss'], penalty=param_fixed['penalty'])
+
+    # joblib is used to cross-validate in parallel by setting `n_jobs=-1` in GridSearchCV
+    # Supress joblib warning. See https://github.com/scikit-learn/scikit-learn/issues/6370
+    warnings.filterwarnings('ignore', message='Changing the shape of non-C contiguous array')
+    clf_grid = grid_search.GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc')
+    pipeline = make_pipeline(
+        feature_select,  # Feature selection
+        StandardScaler(),  # Feature scaling
+        clf_grid)
+    pipeline.fit(X=X_train, y=y_train)
+    
+    y_pred_train = pipeline.decision_function(X_train)
+    y_pred_test = pipeline.decision_function(X_test)
+
+    metrics_train = get_threshold_metrics(y_train, y_pred_train)
+    metrics_test = get_threshold_metrics(y_test, y_pred_test)
+    
+    return metrics_train, metrics_test
+
+
+# Now run on data with and without covariates. We'll drop the numeric covariates, since they contain Nan's that we haven't handled.
+
+# In[14]:
+
+expression_metrics_train, expression_metrics_test = gway_classifier(X, y)
+okay_covariates = covariates_df.drop(numeric_columns, axis=1)
+covariates_metrics_train, covariates_metrics_test = gway_classifier(pd.concat([X, okay_covariates], axis=1), y)
+
+
+# And plot...
+
+# In[15]:
+
+# Plot ROC
+plt.figure()
+labels = ('Without covariates (training)',
+          'Without covariates (testing)',
+          'With covariates (training)', 
+          'With covariates (testing)')
+metrics = (expression_metrics_train,
+           expression_metrics_test,
+           covariates_metrics_train,
+           covariates_metrics_test)
+for label, metric in zip(labels, metrics):
+    roc_df = metric['roc_df']
+    plt.plot(roc_df.fpr, roc_df.tpr,
+        label='{} (AUROC = {:.1%})'.format(label, metric['auroc']))
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('Predicting TP53 mutation from gene expression (ROC curves)')
+plt.legend(loc='lower right');
+
+
+# Unfortunately, the classifier does not even seem to notice the presence of the covariates. But this could be due to the MAD filtering step, lack of feature normalization, and plenty of other things. Curious to hear thoughts about where to go.
```

Comment:
My guess is that the MAD selection is removing the covariates. But even if not, covariates can still be helpful even if they don't improve overall performance -- they can still prevent confounding (faulty feature importance measurements).